File: implementations/ts/src/policy.ts
// Resolution policies and Views (SPEC-5, ERRATA-5). resolve : Policy -> HView -> View is the
// only exit from the algebra into application space; all pluralism is policy choice (P5).

import { type CborValue, array, bool, encode, float, map, tstr } from "./cbor.js";
import { bytesToHex } from "./hash.js";
import type { HVEntry, HView } from "./hview.js";
import { comparePrimitives, evalPred, type Pred } from "./pred.js";
import type { Primitive, Target } from "./types.js";

export type View = Primitive | readonly View[] | { readonly [key: string]: View };

export type MergeFn = "max" | "min" | "sum" | "count" | "and" | "or" | "concatSorted";

export type Order =
  | { readonly kind: "byTimestamp"; readonly dir: "desc" | "asc" }
  | { readonly kind: "byAuthorRank"; readonly authors: readonly string[] }
  | { readonly kind: "byPred"; readonly pred: Pred; readonly then: Order }
  | { readonly kind: "lexById" };

export type PropPolicy =
  | { readonly kind: "pick"; readonly order: Order }
  | { readonly kind: "all"; readonly order: Order }
  | { readonly kind: "merge"; readonly fn: MergeFn }
  | { readonly kind: "conflicts"; readonly order: Order }
  | { readonly kind: "absentAs"; readonly constant: Primitive; readonly then: PropPolicy };

export interface Policy {
  readonly props: ReadonlyMap<string, PropPolicy>;
  readonly default: PropPolicy;
}

// --- ordering (R3: every chain ends in an implicit lexById tiebreak) ------------------------------

function cmpByOrder(order: Order, a: HVEntry, b: HVEntry): number {
  switch (order.kind) {
    case "byTimestamp": {
      const d = a.delta.claims.timestamp - b.delta.claims.timestamp;
      if (d !== 0) return order.dir === "desc" ? -d : d;
      return 0;
    }
    case "byAuthorRank": {
      const rank = (author: string) => {
        const i = order.authors.indexOf(author);
        return i === -1 ? order.authors.length : i;
      };
      return rank(a.delta.claims.author) - rank(b.delta.claims.author);
    }
    case "byPred": {
      const am = evalPred(order.pred, a.delta) ? 0 : 1;
      const bm = evalPred(order.pred, b.delta) ? 0 : 1;
      if (am !== bm) return am - bm; // matches first
      return cmpByOrder(order.then, a, b);
    }
    case "lexById":
      return a.delta.id < b.delta.id ? -1 : a.delta.id > b.delta.id ? 1 : 0;
  }
}

function sortEntries(order: Order, entries: readonly HVEntry[]): HVEntry[] {
  return [...entries].sort((a, b) => {
    const primary = cmpByOrder(order, a, b);
    if (primary !== 0) return primary;
    return a.delta.id < b.delta.id ? -1 : a.delta.id > b.delta.id ? 1 : 0;
  });
}

// --- candidate value extraction (R1) ---------------------------------------------------------------

function renderTarget(t: Target, expansion: HView | undefined, policy: Policy): View {
  if (expansion !== undefined) return resolveView(policy, expansion);
  switch (t.kind) {
    case "primitive":
      return t.value;
    case "entity":
      return t.entity.id;
    case "delta":
      return t.deltaRef.delta;
  }
}

function candidateValue(e: HVEntry, root: string, policy: Policy): View {
  const nonFiling: Array<[string, View]> = [];
  e.delta.claims.pointers.forEach((p, i) => {
    const filing = p.target.kind === "entity" && p.target.entity.id === root;
    if (filing) return;
    nonFiling.push([p.role, renderTarget(p.target, e.expanded?.get(i), policy)]);
  });
  if (nonFiling.length === 0) return true; // the bare fact of the edge
  if (nonFiling.length === 1) return nonFiling[0]![1];
  const obj: Record<string, View> = {};
  for (const [role, v] of nonFiling) {
    const existing = obj[role];
    if (existing === undefined) obj[role] = v;
    else if (Array.isArray(existing)) obj[role] = [...existing, v];
    else obj[role] = [existing, v];
  }
  return obj;
}

// --- View canonical form (R4) ----------------------------------------------------------------------

export function viewToCbor(v: View): CborValue {
  if (typeof v === "string") return tstr(v);
  if (typeof v === "number") return float(v);
  if (typeof v === "boolean") return bool(v);
  if (Array.isArray(v)) return array(v.map(viewToCbor));
  const entries = Object.entries(v as { [key: string]: View }).map(
    ([k, x]): readonly [string, CborValue] => [k, viewToCbor(x)],
  );
  return map(entries);
}

export function viewCanonicalHex(v: View): string {
  return bytesToHex(encode(viewToCbor(v)));
}

// --- resolution ------------------------------------------------------------------------------------

const ABSENT = Symbol("absent");
type Resolved = View | typeof ABSENT;

function isPrimitive(v: View): v is Primitive {
  return typeof v === "string" || typeof v === "number" || typeof v === "boolean";
}

function applyMerge(
  fn: MergeFn,
  entries: readonly HVEntry[],
  root: string,
  policy: Policy,
): Resolved {
  // Fold in ascending delta-id order — float addition is order-dependent (R2).
  const sorted = sortEntries({ kind: "lexById" }, entries);
  if (fn === "count") return sorted.length === 0 ? ABSENT : sorted.length;
  const prims = sorted
    .map((e) => candidateValue(e, root, policy))
    .filter((v): v is Primitive => isPrimitive(v));
  switch (fn) {
    case "max":
    case "min": {
      if (prims.length === 0) return ABSENT;
      return prims.reduce((acc, v) => {
        const c = comparePrimitives(v, acc);
        return (fn === "max" ? c > 0 : c < 0) ? v : acc;
      });
    }
    case "sum": {
      const nums = prims.filter((v): v is number => typeof v === "number");
      if (nums.length === 0) return ABSENT;
      return nums.reduce((a, b) => a + b, 0);
    }
    case "and":
    case "or": {
      const bools = prims.filter((v): v is boolean => typeof v === "boolean");
      if (bools.length === 0) return ABSENT;
      return fn === "and" ? bools.every(Boolean) : bools.some(Boolean);
    }
    case "concatSorted": {
      if (prims.length === 0) return ABSENT;
      return [...prims].sort(comparePrimitives);
    }
  }
}

function applyPropPolicy(
  pp: PropPolicy,
  entries: readonly HVEntry[],
  root: string,
  policy: Policy,
): Resolved {
  switch (pp.kind) {
    case "pick": {
      if (entries.length === 0) return ABSENT;
      const sorted = sortEntries(pp.order, entries);
      return candidateValue(sorted[0]!, root, policy);
    }
    case "all": {
      if (entries.length === 0) return ABSENT;
      return sortEntries(pp.order, entries).map((e) => candidateValue(e, root, policy));
    }
    case "merge":
      return applyMerge(pp.fn, entries, root, policy);
    case "conflicts": {
      const sorted = sortEntries(pp.order, entries);
      const seen = new Set<string>();
      const distinct: View[] = [];
      for (const e of sorted) {
        const v = candidateValue(e, root, policy);
        const key = viewCanonicalHex(v);
        if (!seen.has(key)) {
          seen.add(key);
          distinct.push(v);
        }
      }
      return distinct.length >= 2 ? distinct : ABSENT;
    }
    case "absentAs": {
      const inner = applyPropPolicy(pp.then, entries, root, policy);
      return inner === ABSENT ? pp.constant : inner;
    }
  }
}

// resolve(policy, HView) -> View. Deterministic; total; provenance-optional (SPEC-5 §2).
// The View covers every property named in the policy plus every HView property (R3).
export function resolveView(policy: Policy, hview: HView): View {
  const keys = new Set<string>([...policy.props.keys(), ...hview.props.keys()]);
  const obj: Record<string, View> = {};
  for (const key of keys) {
    const entries = hview.props.get(key) ?? [];
    const pp = policy.props.get(key) ?? policy.default;
    const v = applyPropPolicy(pp, entries, hview.id, policy);
    if (v !== ABSENT) obj[key] = v;
  }
  return obj;
}

File: implementations/rust/src/policy.rs
//! Resolution policies and Views (SPEC-5, ERRATA-5). Mirrors ../ts/src/policy.ts.

use std::cmp::Ordering;
use std::collections::{BTreeMap, BTreeSet};

use crate::cbor::{encode, CborValue};
use crate::hview::{HVEntry, HView};
use crate::pred::{compare_primitives, eval_pred, Pred};
use crate::types::{Primitive, Target};

#[derive(Debug, Clone, PartialEq)]
pub enum View {
    Prim(Primitive),
    Arr(Vec<View>),
    Obj(BTreeMap<String, View>),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MergeFn {
    Max,
    Min,
    Sum,
    Count,
    And,
    Or,
    ConcatSorted,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Order {
    ByTimestamp { desc: bool },
    ByAuthorRank(Vec<String>),
    ByPred { pred: Pred, then: Box<Order> },
    LexById,
}

#[derive(Debug, Clone, PartialEq)]
pub enum PropPolicy {
    Pick(Order),
    All(Order),
    Merge(MergeFn),
    Conflicts(Order),
    AbsentAs {
        constant: Primitive,
        then: Box<PropPolicy>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub struct Policy {
    pub props: BTreeMap<String, PropPolicy>,
    pub default: PropPolicy,
}

// --- ordering (R3: every chain ends in an implicit lexById tiebreak) ------------------------------

fn cmp_by_order(order: &Order, a: &HVEntry, b: &HVEntry) -> Ordering {
    match order {
        Order::ByTimestamp { desc } => {
            let o = a
                .delta
                .claims
                .timestamp
                .partial_cmp(&b.delta.claims.timestamp)
                .expect("timestamps are finite");
            if *desc {
                o.reverse()
            } else {
                o
            }
        }
        Order::ByAuthorRank(authors) => {
            let rank = |author: &str| {
                authors
                    .iter()
                    .position(|a| a == author)
                    .unwrap_or(authors.len())
            };
            rank(&a.delta.claims.author).cmp(&rank(&b.delta.claims.author))
        }
        Order::ByPred { pred, then } => {
            let am = !eval_pred(pred, &a.delta, None);
            let bm = !eval_pred(pred, &b.delta, None);
            match am.cmp(&bm) {
                Ordering::Equal => cmp_by_order(then, a, b),
                o => o, // matches (false < true after negation) first
            }
        }
        Order::LexById => a.delta.id.cmp(&b.delta.id),
    }
}

fn sort_entries<'a>(order: &Order, entries: &'a [HVEntry]) -> Vec<&'a HVEntry> {
    let mut out: Vec<&HVEntry> = entries.iter().collect();
    out.sort_by(|a, b| cmp_by_order(order, a, b).then_with(|| a.delta.id.cmp(&b.delta.id)));
    out
}

// --- candidate value extraction (R1) ---------------------------------------------------------------

fn render_target(t: &Target, expansion: Option<&HView>, policy: &Policy) -> View {
    if let Some(h) = expansion {
        return resolve_view(policy, h);
    }
    match t {
        Target::Primitive(p) => View::Prim(p.clone()),
        Target::Entity(e) => View::Prim(Primitive::Str(e.id.clone())),
        Target::Delta(d) => View::Prim(Primitive::Str(d.delta.clone())),
    }
}

fn candidate_value(e: &HVEntry, root: &str, policy: &Policy) -> View {
    let mut non_filing: Vec<(String, View)> = Vec::new();
    for (i, p) in e.delta.claims.pointers.iter().enumerate() {
        let filing = matches!(&p.target, Target::Entity(er) if er.id == root);
        if filing {
            continue;
        }
        non_filing.push((
            p.role.clone(),
            render_target(&p.target, e.expanded.get(&i), policy),
        ));
    }
    if non_filing.is_empty() {
        return View::Prim(Primitive::Bool(true)); // the bare fact of the edge
    }
    if non_filing.len() == 1 {
        return non_filing.into_iter().next().unwrap().1;
    }
    let mut obj: BTreeMap<String, View> = BTreeMap::new();
    for (role, v) in non_filing {
        match obj.remove(&role) {
            None => {
                obj.insert(role, v);
            }
            Some(View::Arr(mut xs)) => {
                xs.push(v);
                obj.insert(role, View::Arr(xs));
            }
            Some(existing) => {
                obj.insert(role, View::Arr(vec![existing, v]));
            }
        }
    }
    View::Obj(obj)
}

// --- View canonical form (R4) ----------------------------------------------------------------------

pub fn view_to_cbor(v: &View) -> CborValue {
    match v {
        View::Prim(Primitive::Str(s)) => CborValue::Tstr(s.clone()),
        View::Prim(Primitive::Num(n)) => CborValue::Float(*n),
        View::Prim(Primitive::Bool(b)) => CborValue::Bool(*b),
        View::Arr(xs) => CborValue::Array(xs.iter().map(view_to_cbor).collect()),
        View::Obj(m) => CborValue::Map(
            m.iter()
                .map(|(k, x)| (k.clone(), view_to_cbor(x)))
                .collect(),
        ),
    }
}

pub fn view_canonical_hex(v: &View) -> String {
    hex::encode(encode(&view_to_cbor(v)))
}

// --- resolution ------------------------------------------------------------------------------------

fn is_primitive(v: &View) -> Option<&Primitive> {
    match v {
        View::Prim(p) => Some(p),
        _ => None,
    }
}

fn apply_merge(fn_: MergeFn, entries: &[HVEntry], root: &str, policy: &Policy) -> Option<View> {
    // Fold in ascending delta-id order — float addition is order-dependent (R2).
    let sorted = sort_entries(&Order::LexById, entries);
    if fn_ == MergeFn::Count {
        return if sorted.is_empty() {
            None
        } else {
            Some(View::Prim(Primitive::Num(sorted.len() as f64)))
        };
    }
    let prims: Vec<Primitive> = sorted
        .iter()
        .map(|e| candidate_value(e, root, policy))
        .filter_map(|v| is_primitive(&v).cloned())
        .collect();
    match fn_ {
        MergeFn::Max | MergeFn::Min => prims
            .into_iter()
            .reduce(|acc, v| {
                let c = compare_primitives(&v, &acc);
                let take = if fn_ == MergeFn::Max {
                    c == Ordering::Greater
                } else {
                    c == Ordering::Less
                };
                if take {
                    v
                } else {
                    acc
                }
            })
            .map(View::Prim),
        MergeFn::Sum => {
            let nums: Vec<f64> = prims
                .iter()
                .filter_map(|p| match p {
                    Primitive::Num(n) => Some(*n),
                    _ => None,
                })
                .collect();
            if nums.is_empty() {
                None
            } else {
                Some(View::Prim(Primitive::Num(
                    nums.iter().fold(0.0, |a, b| a + b),
                )))
            }
        }
        MergeFn::And | MergeFn::Or => {
            let bools: Vec<bool> = prims
                .iter()
                .filter_map(|p| match p {
                    Primitive::Bool(b) => Some(*b),
                    _ => None,
                })
                .collect();
            if bools.is_empty() {
                None
            } else if fn_ == MergeFn::And {
                Some(View::Prim(Primitive::Bool(bools.iter().all(|b| *b))))
            } else {
                Some(View::Prim(Primitive::Bool(bools.iter().any(|b| *b))))
            }
        }
        MergeFn::ConcatSorted => {
            if prims.is_empty() {
                None
            } else {
                let mut sorted_prims = prims;
                sorted_prims.sort_by(compare_primitives);
                Some(View::Arr(
                    sorted_prims.into_iter().map(View::Prim).collect(),
                ))
            }
        }
        MergeFn::Count => unreachable!("handled above"),
    }
}

fn apply_prop_policy(
    pp: &PropPolicy,
    entries: &[HVEntry],
    root: &str,
    policy: &Policy,
) -> Option<View> {
    match pp {
        PropPolicy::Pick(order) => {
            if entries.is_empty() {
                return None;
            }
            let sorted = sort_entries(order, entries);
            Some(candidate_value(sorted[0], root, policy))
        }
        PropPolicy::All(order) => {
            if entries.is_empty() {
                return None;
            }
            Some(View::Arr(
                sort_entries(order, entries)
                    .iter()
                    .map(|e| candidate_value(e, root, policy))
                    .collect(),
            ))
        }
        PropPolicy::Merge(fn_) => apply_merge(*fn_, entries, root, policy),
        PropPolicy::Conflicts(order) => {
            let sorted = sort_entries(order, entries);
            let mut seen: BTreeSet<String> = BTreeSet::new();
            let mut distinct: Vec<View> = Vec::new();
            for e in sorted {
                let v = candidate_value(e, root, policy);
                if seen.insert(view_canonical_hex(&v)) {
                    distinct.push(v);
                }
            }
            if distinct.len() >= 2 {
                Some(View::Arr(distinct))
            } else {
                None
            }
        }
        PropPolicy::AbsentAs { constant, then } => apply_prop_policy(then, entries, root, policy)
            .or_else(|| Some(View::Prim(constant.clone()))),
    }
}

/// resolve(policy, HView) -> View. Deterministic; total; provenance-optional (SPEC-5 §2).
/// The View covers every property named in the policy plus every HView property (R3).
pub fn resolve_view(policy: &Policy, hview: &HView) -> View {
    let mut keys: BTreeSet<&String> = policy.props.keys().collect();
    keys.extend(hview.props.keys());
    let empty: Vec<HVEntry> = Vec::new();
    let mut obj: BTreeMap<String, View> = BTreeMap::new();
    for key in keys {
        let entries = hview.props.get(key).unwrap_or(&empty);
        let pp = policy.props.get(key).unwrap_or(&policy.default);
        if let Some(v) = apply_prop_policy(pp, entries, &hview.id, policy) {
            obj.insert(key.clone(), v);
        }
    }
    View::Obj(obj)
}
